File: Classes/View Controllers/Tabs/Playlists/PlaylistSongsViewController.m
//
//  PlaylistSongsViewController.m
//  iSub
//
//  Created by Ben Baron on 4/2/10.
//  Copyright 2010 Ben Baron. All rights reserved.
//

#import "PlaylistSongsViewController.h"
#import "ServerListViewController.h"
#import "UIViewController+PushViewControllerCustom.h"
#import "NSMutableURLRequest+SUS.h"
#import "ViewObjectsSingleton.h"
#import "Defines.h"
#import "RXMLElement.h"
#import "FMDatabaseQueueAdditions.h"
#import "SavedSettings.h"
#import "PlaylistSingleton.h"
#import "MusicSingleton.h"
#import "DatabaseSingleton.h"
#import "JukeboxSingleton.h"
#import "NSError+ISMSError.h"
#import "ISMSSong+DAO.h"
#import "SUSServerPlaylist.h"
#import "EX2Kit.h"
#import "Swift.h"
#import "SUSLoader.h"

LOG_LEVEL_ISUB_DEFAULT

@interface PlaylistSongsViewController()
@property (strong) NSURLSessionDataTask *dataTask;
@property (strong) UIView *uploadBannerView;
@end

@implementation PlaylistSongsViewController

- (BOOL)isLocalPlaylist {
    return self.serverPlaylist == nil;
}

- (void)viewDidLoad {
    [super viewDidLoad];

    if (self.isLocalPlaylist) {
		self.title = [databaseS.localPlaylistsDbQueue stringForQuery:@"SELECT playlist FROM localPlaylists WHERE md5 = ?", self.md5];

		if (!settingsS.isOfflineMode) {
			UIView *uploadView = [[UIView alloc] initWithFrame:CGRectMake(0, 0, 320, 50)];

			UILabel *sendLabel = [[UILabel alloc] initWithFrame:CGRectMake(0, 0, 320, 50)];
			sendLabel.autoresizingMask = UIViewAutoresizingFlexibleWidth | UIViewAutoresizingFlexibleRightMargin;
			sendLabel.backgroundColor = [UIColor clearColor];
            sendLabel.textColor = UIColor.labelColor;
			sendLabel.textAlignment = NSTextAlignmentCenter;
			sendLabel.font = [UIFont boldSystemFontOfSize:24];
			sendLabel.text = @"Save to Server";
			[uploadView addSubview:sendLabel];

			UIButton *sendButton = [UIButton buttonWithType:UIButtonTypeCustom];
			sendButton.autoresizingMask = UIViewAutoresizingFlexibleWidth | UIViewAutoresizingFlexibleRightMargin;
			sendButton.frame = CGRectMake(0, 0, 320, 50);
			[sendButton addTarget:self action:@selector(uploadPlaylistAction:) forControlEvents:UIControlEventTouchUpInside];
			[uploadView addSubview:sendButton];

			self.uploadBannerView = uploadView;
		}
	} else {
        self.title = self.serverPlaylist.playlistName;
        self.playlistCount = [databaseS.localPlaylistsDbQueue intForQuery:[NSString stringWithFormat:@"SELECT COUNT(*) FROM splaylist%@", self.md5]];
        [self updatePlaylistHeader];
		[self.tableView reloadData];

        // Add the pull to refresh view
        __weak PlaylistSongsViewController *weakSelf = self;
        self.refreshControl = [[RefreshControl alloc] initWithHandler:^{
            [weakSelf loadData];
        }];
	}

    self.tableView.rowHeight = Defines.rowHeight;
    [self.tableView registerClass:UniversalTableViewCell.class forCellReuseIdentifier:UniversalTableViewCell.reuseId];
}

- (void)updatePlaylistHeader {
    NSString *name = self.title ?: @"";
    NSUInteger count = self.playlistCount;

    // Collect up to 4 cover art IDs from the first 4 songs
    NSMutableArray<NSString *> *artIds = [NSMutableArray arrayWithCapacity:4];
    NSString *tableName = [NSString stringWithFormat:@"%@%@", self.isLocalPlaylist ? @"playlist" : @"splaylist", self.md5];
    for (NSUInteger i = 0; i < MIN(count, 4); i++) {
        ISMSSong *song = self.isLocalPlaylist
            ? [ISMSSong songFromDbRow:i inTable:tableName inDatabaseQueue:databaseS.localPlaylistsDbQueue]
            : [ISMSSong songFromServerPlaylistId:self.md5 row:i];
        if (song.coverArtId) {
            [artIds addObject:song.coverArtId];
        }
    }

    __weak PlaylistSongsViewController *weakSelf = self;

    PlaylistHeaderView *header = [[PlaylistHeaderView alloc] init];
    header.playlistName = name;
    header.songCount = (NSInteger)count;
    header.coverArtIds = artIds;

    header.onPlayAll = ^{
        [weakSelf playAllSongs];
    };
    header.onShuffle = ^{
        [weakSelf shuffleSongs];
    };

    // Build a container that also includes the upload banner for local playlists
    UIView *container = [[UIView alloc] init];
    container.translatesAutoresizingMaskIntoConstraints = NO;
    self.tableView.tableHeaderView = container;
    [NSLayoutConstraint activateConstraints:@[
        [container.centerXAnchor constraintEqualToAnchor:self.tableView.centerXAnchor],
        [container.widthAnchor constraintEqualToAnchor:self.tableView.widthAnchor],
        [container.topAnchor constraintEqualToAnchor:self.tableView.topAnchor]
    ]];

    [container addSubview:header];
    header.translatesAutoresizingMaskIntoConstraints = NO;
    [NSLayoutConstraint activateConstraints:@[
        [header.leadingAnchor constraintEqualToAnchor:container.leadingAnchor],
        [header.trailingAnchor constraintEqualToAnchor:container.trailingAnchor],
        [header.topAnchor constraintEqualToAnchor:container.topAnchor],
        [header.heightAnchor constraintEqualToConstant:PlaylistHeaderView.height]
    ]];

    UIView *bottomAnchorView = header;
    if (self.uploadBannerView) {
        UIView *uploadBanner = self.uploadBannerView;
        [container addSubview:uploadBanner];
        uploadBanner.translatesAutoresizingMaskIntoConstraints = NO;
        [NSLayoutConstraint activateConstraints:@[
            [uploadBanner.leadingAnchor constraintEqualToAnchor:container.leadingAnchor],
            [uploadBanner.trailingAnchor constraintEqualToAnchor:container.trailingAnchor],
            [uploadBanner.topAnchor constraintEqualToAnchor:header.bottomAnchor],
            [uploadBanner.heightAnchor constraintEqualToConstant:50]
        ]];
        bottomAnchorView = uploadBanner;
    }

    [bottomAnchorView.bottomAnchor constraintEqualToAnchor:container.bottomAnchor].active = YES;

    [self.tableView.tableHeaderView layoutIfNeeded];
    self.tableView.tableHeaderView = self.tableView.tableHeaderView;
}

- (void)playAllSongs {
    if (settingsS.isJukeboxEnabled) {
        [databaseS resetJukeboxPlaylist];
        [jukeboxS clearRemotePlaylist];
    } else {
        [databaseS resetCurrentPlaylistDb];
    }
    playlistS.isShuffle = NO;
    [self loadPlaylistIntoCurrentPlaylist];
    if (settingsS.isJukeboxEnabled) {
        [jukeboxS replacePlaylistWithLocal];
    }
    [musicS playSongAtPosition:0];
}

- (void)shuffleSongs {
    if (settingsS.isJukeboxEnabled) {
        [databaseS resetJukeboxPlaylist];
        [jukeboxS clearRemotePlaylist];
    } else {
        [databaseS resetCurrentPlaylistDb];
    }
    playlistS.isShuffle = YES;
    [self loadPlaylistIntoCurrentPlaylist];
    if (settingsS.isJukeboxEnabled) {
        [jukeboxS replacePlaylistWithLocal];
    }
    [musicS playSongAtPosition:0];
}

- (void)loadPlaylistIntoCurrentPlaylist {
    NSString *databaseName = settingsS.isOfflineMode ? @"offlineCurrentPlaylist.db" : [NSString stringWithFormat:@"%@currentPlaylist.db", [settingsS.urlString md5]];
    NSString *currTableName = settingsS.isJukeboxEnabled ? @"jukeboxCurrentPlaylist" : @"currentPlaylist";
    NSString *playTableName = [NSString stringWithFormat:@"%@%@", self.isLocalPlaylist ? @"playlist" : @"splaylist", self.md5];
    [databaseS.localPlaylistsDbQueue inDatabase:^(FMDatabase *db) {
        [db executeUpdate:@"ATTACH DATABASE ? AS ?", [databaseS.databaseFolderPath stringByAppendingPathComponent:databaseName], @"currentPlaylistDb"];
        if ([db hadError]) { DDLogError(@"[PlaylistSongsViewController] Err attaching the currentPlaylistDb %d: %@", [db lastErrorCode], [db lastErrorMessage]); }
        [db executeUpdate:[NSString stringWithFormat:@"INSERT INTO %@ SELECT * FROM %@", currTableName, playTableName]];
        [db executeUpdate:@"DETACH DATABASE currentPlaylistDb"];
    }];
}

- (void)loadData {
    NSDictionary *parameters = [NSDictionary dictionaryWithObject:n2N(self.serverPlaylist.playlistId) forKey:@"id"];
    NSMutableURLRequest *request = [NSMutableURLRequest requestWithSUSAction:@"getPlaylist" parameters:parameters];
    self.dataTask = [SUSLoader.sharedSession dataTaskWithRequest:request completionHandler:^(NSData * _Nullable data, NSURLResponse * _Nullable response, NSError * _Nullable error) {
        if (error) {
            [EX2Dispatch runInMainThreadAsync:^{
                if (settingsS.isPopupsEnabled) {
                    NSString *message = [NSString stringWithFormat:@"There was an error loading the playlist.\n\nError %li: %@", (long)error.code, error.localizedDescription];
                    UIAlertController *alert = [UIAlertController alertControllerWithTitle:@"Error" message:message preferredStyle:UIAlertControllerStyleAlert];
                    [alert addAction:[UIAlertAction actionWithTitle:@"OK" style:UIAlertActionStyleCancel handler:nil]];
                    [self presentViewController:alert animated:YES completion:nil];
                }
                
                self.tableView.scrollEnabled = YES;
                [viewObjectsS hideLoadingScreen];
                [self.refreshControl endRefreshing];
            }];
        } else {
            RXMLElement *root = [[RXMLElement alloc] initFromXMLData:data];
            if (!root.isValid) {
                //NSError *error = [NSError errorWithISMSCode:ISMSErrorCode_NotXML];
                // TODO: Handle this error
            } else {
                RXMLElement *error = [root child:@"error"];
                if (error.isValid) {
                    //NSString *code = [error attribute:@"code"];
                    //NSString *message = [error attribute:@"message"];
                    //[self subsonicErrorCode:[code intValue] message:message];
                    // TODO: Handle this error
                } else {
                    // TODO: Handle !isValid case
                    if ([[root child:@"playlist"] isValid]) {
                        [databaseS removeServerPlaylistTable:self.md5];
                        [databaseS createServerPlaylistTable:self.md5];
                        [root iterate:@"playlist.entry" usingBlock:^(RXMLElement *e) {
                            ISMSSong *aSong = [[ISMSSong alloc] initWithRXMLElement:e];
                            [aSong insertIntoServerPlaylistWithPlaylistId:self.md5];
                        }];
                    }
                }
            }
            
            self.playlistCount = [databaseS.localPlaylistsDbQueue intForQuery:[NSString stringWithFormat:@"SELECT COUNT(*) FROM splaylist%@", self.md5]];

            [EX2Dispatch runInMainThreadAsync:^{
                [self updatePlaylistHeader];
                [self.tableView reloadData];
                [self.refreshControl endRefreshing];
                [viewObjectsS hideLoadingScreen];
                self.tableView.scrollEnabled = YES;
            }];
        }
    }];
    [self.dataTask resume];
    
    self.tableView.scrollEnabled = NO;
    [viewObjectsS showAlbumLoadingScreen:self.view sender:self];
}	

- (void)cancelLoad {
    [self.dataTask cancel];
    self.dataTask = nil;
	self.tableView.scrollEnabled = YES;
	[viewObjectsS hideLoadingScreen];
	
	if (!self.isLocalPlaylist) {
        [self.refreshControl endRefreshing];
	}
}

- (void)viewWillAppear:(BOOL)animated  {
    [super viewWillAppear:animated];
    
    // For some reason this controller needs to do this, but none of the others do :/
//    self.navigationController.navigationBar.translucent = NO;
	
	if(musicS.showPlayerIcon) {
		self.navigationItem.rightBarButtonItem = [[UIBarButtonItem alloc] initWithImage:[UIImage systemImageNamed:Defines.musicNoteImageSystemName] style:UIBarButtonItemStylePlain target:self action:@selector(nowPlayingAction:)];
	} else {
		self.navigationItem.rightBarButtonItem = nil;
	}
	
	if (self.isLocalPlaylist) {
		self.playlistCount = [databaseS.localPlaylistsDbQueue intForQuery:[NSString stringWithFormat:@"SELECT COUNT(*) FROM playlist%@", self.md5]];
        [self updatePlaylistHeader];
		[self.tableView reloadData];
	} else {
		if (self.playlistCount == 0) {
			[self loadData];
		}
	}
}

- (void) settingsAction:(id)sender  {
	ServerListViewController *serverListViewController = [[ServerListViewController alloc] initWithNibName:@"ServerListViewController" bundle:nil];
	serverListViewController.hidesBottomBarWhenPushed = YES;
	[self.navigationController pushViewController:serverListViewController animated:YES];
}


- (IBAction)nowPlayingAction:(id)sender {
    PlayerViewController *playerViewController = [[PlayerViewController alloc] init];
    playerViewController.hidesBottomBarWhenPushed = YES;
	[self.navigationController pushViewController:playerViewController animated:YES];
}

- (void)uploadPlaylistAction:(id)sender {
	NSMutableDictionary *parameters = [NSMutableDictionary dictionaryWithObjectsAndKeys:n2N(self.title), @"name", nil];
    
	NSString *query = [NSString stringWithFormat:@"SELECT COUNT(*) FROM playlist%@", self.md5];
	NSUInteger count = [databaseS.localPlaylistsDbQueue intForQuery:query];
	NSMutableArray *songIds = [NSMutableArray arrayWithCapacity:count];
	for (int i = 1; i <= count; i++) {
		@autoreleasepool {
			NSString *query = [NSString stringWithFormat:@"SELECT songId FROM playlist%@ WHERE ROWID = %i", self.md5, i];
			NSString *songId = [databaseS.localPlaylistsDbQueue stringForQuery:query];
			
			[songIds addObject:n2N(songId)];
		}
	}
	[parameters setObject:[NSArray arrayWithArray:songIds] forKey:@"songId"];
	
    NSMutableURLRequest *request = [NSMutableURLRequest requestWithSUSAction:@"createPlaylist" parameters:parameters];
    self.dataTask = [SUSLoader.sharedSession dataTaskWithRequest:request completionHandler:^(NSData * _Nullable data, NSURLResponse * _Nullable response, NSError * _Nullable error) {
        if (error) {
            if (settingsS.isPopupsEnabled) {
                [EX2Dispatch runInMainThreadAsync:^{
                    NSString *message = [NSString stringWithFormat:@"There was an error saving the playlist to the server.\n\nError %li: %@", (long)error.code, error.localizedDescription];
                    UIAlertController *alert = [UIAlertController alertControllerWithTitle:@"Error" message:message preferredStyle:UIAlertControllerStyleAlert];
                    [alert addAction:[UIAlertAction actionWithTitle:@"OK" style:UIAlertActionStyleCancel handler:nil]];
                    [self presentViewController:alert animated:YES completion:nil];
                }];
            }
        } else {
            DDLogVerbose(@"[PlaylistSongsViewController] upload playlist response: %@", [[NSString alloc] initWithData:data encoding:NSUTF8StringEncoding]);
            RXMLElement *root = [[RXMLElement alloc] initFromXMLData:data];
            if (!root.isValid) {
                NSError *error = [NSError errorWithISMSCode:ISMSErrorCode_NotXML];
                [self subsonicErrorCode:nil message:error.description];
            } else {
                RXMLElement *error = [root child:@"error"];
                if (error.isValid) {
                    NSString *code = [error attribute:@"code"];
                    NSString *message = [error attribute:@"message"];
                    [self subsonicErrorCode:code message:message];
                }
            }
        }
        
        [EX2Dispatch runInMainThreadAsync:^{
            self.tableView.scrollEnabled = YES;
            [viewObjectsS hideLoadingScreen];
            [self.refreshControl endRefreshing];
        }];
    }];
    [self.dataTask resume];
    
    self.tableView.scrollEnabled = NO;
    [viewObjectsS showAlbumLoadingScreen:self.view sender:self];
}

- (void)subsonicErrorCode:(NSString *)errorCode message:(NSString *)message {
    DDLogError(@"[PlayistSongsViewController] subsonic error %@: %@", errorCode, message);
    if (settingsS.isPopupsEnabled) {
        [EX2Dispatch runInMainThreadAsync:^{
            UIAlertController *alert = [UIAlertController alertControllerWithTitle:@"Subsonic Error" message:message preferredStyle:UIAlertControllerStyleAlert];
            [alert addAction:[UIAlertAction actionWithTitle:@"OK" style:UIAlertActionStyleCancel handler:nil]];
            [self presentViewController:alert animated:YES completion:nil];
        }];
    }
}

#pragma mark Table view methods

- (ISMSSong *)songAtIndexPath:(NSIndexPath *)indexPath {
    if (self.isLocalPlaylist) {
        return [ISMSSong songFromDbRow:indexPath.row inTable:[NSString stringWithFormat:@"playlist%@", self.md5] inDatabaseQueue:databaseS.localPlaylistsDbQueue];
    } else {
        return [ISMSSong songFromServerPlaylistId:self.md5 row:indexPath.row];
    }
}

- (NSInteger)numberOfSectionsInTableView:(UITableView *)tableView {
    return 1;
}


// Customize the number of rows in the table view.
- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section {
	return self.playlistCount;
}


// Customize the appearance of table view cells.
- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath {
    UniversalTableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:UniversalTableViewCell.reuseId];
    cell.hideNumberLabel = NO;
    cell.hideCoverArt = NO;
    cell.hideDurationLabel = NO;
    cell.hideSecondaryLabel = NO;
    cell.number = indexPath.row + 1;
    ISMSSong *song = [self songAtIndexPath:indexPath];
    [cell updateWithModel:song];
    if (!song.isVideo) {
        [cell configureSongContextMenuWithSong:song presenter:self];
    }
    return cell;
}

- (void)didSelectRowInternal:(NSIndexPath *)indexPath {
	// Clear the current playlist
	if (settingsS.isJukeboxEnabled) {
		[databaseS resetJukeboxPlaylist];
		[jukeboxS clearRemotePlaylist];
	} else {
		[databaseS resetCurrentPlaylistDb];
	}
	
	playlistS.isShuffle = NO;
	
	/*for (int i = 0; i < self.playlistCount; i++)
	{
		@autoreleasepool
		{
			ISMSSong *aSong;
			if (self.isLocalPlaylist)
			{
				aSong = [ISMSSong songFromDbRow:i inTable:[NSString stringWithFormat:@"playlist%@", self.md5] inDatabaseQueue:databaseS.localPlaylistsDbQueue];
			}
			else
			{
				aSong = [ISMSSong songFromServerPlaylistId:self.md5 row:i];
			}
			
			[aSong addToCurrentPlaylistDbQueue];
		}
	}*/
	
	// Need to do this for speed (NOTE: haha well 10 years ago maybe, but probably not now)
	NSString *databaseName = settingsS.isOfflineMode ? @"offlineCurrentPlaylist.db" : [NSString stringWithFormat:@"%@currentPlaylist.db", [settingsS.urlString md5]];
	NSString *currTableName = settingsS.isJukeboxEnabled ? @"jukeboxCurrentPlaylist" : @"currentPlaylist";
	NSString *playTableName = [NSString stringWithFormat:@"%@%@", self.isLocalPlaylist ? @"playlist" : @"splaylist", self.md5];
	[databaseS.localPlaylistsDbQueue inDatabase:^(FMDatabase *db) {
		 [db executeUpdate:@"ATTACH DATABASE ? AS ?", [databaseS.databaseFolderPath stringByAppendingPathComponent:databaseName], @"currentPlaylistDb"];
		 if ([db hadError]) { DDLogError(@"[PlaylistSongsViewController] Err attaching the currentPlaylistDb %d: %@", [db lastErrorCode], [db lastErrorMessage]); }
		 
		 [db executeUpdate:[NSString stringWithFormat:@"INSERT INTO %@ SELECT * FROM %@", currTableName, playTableName]];
		 [db executeUpdate:@"DETACH DATABASE currentPlaylistDb"];
	 }];
	
    if (settingsS.isJukeboxEnabled) {
		[jukeboxS replacePlaylistWithLocal];
    }

    [viewObjectsS hideLoadingScreen];
    
    [musicS playSongAtPosition:indexPath.row];
}


- (void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath  {
	if (!indexPath) return;
    
    [viewObjectsS showLoadingScreenOnMainWindowWithMessage:nil];
    [self performSelector:@selector(didSelectRowInternal:) withObject:indexPath afterDelay:0.05];
}

- (UISwipeActionsConfiguration *)tableView:(UITableView *)tableView trailingSwipeActionsConfigurationForRowAtIndexPath:(NSIndexPath *)indexPath {
    ISMSSong *song = [self songAtIndexPath:indexPath];
    if (!song.isVideo) {
        return [SwipeAction downloadAndQueueConfigWithModel:song];
    }
    return nil;
}

@end

File: Classes/View Controllers/Tabs/Playlists/PlaylistSongsViewController.swift
//
//  PlaylistSongsViewController.swift
//  iSub
//
//  Created by François ND on 2026-03-23.
//  Copyright © 2026 François ND. All rights reserved.
//

import UIKit
import CocoaLumberjackSwift

// MARK: - PlaylistSource

private enum PlaylistSource {
    case local(md5: String, count: UInt)
    case server(ServerPlaylist)
}

// MARK: - PlaylistSongsViewController

@objc(PlaylistSongsViewController) final class PlaylistSongsViewController: UITableViewController {

    // MARK: - ObjC-visible properties (callers set these before pushing)

    @objc var md5: String = ""
    @objc var playlistCount: UInt = 0
    @objc var serverPlaylist: ServerPlaylist? = nil

    // MARK: - Private state

    private var source: PlaylistSource = .local(md5: "", count: 0)
    private var songs: [Song] = []
    private var fetchTask: Task<Void, Never>?

    // MARK: - Lifecycle

    override func viewDidLoad() {
        super.viewDidLoad()

        // Resolve source from ObjC-set properties
        if let sp = serverPlaylist {
            source = .server(sp)
            title = sp.playlistName
        } else {
            source = .local(md5: md5, count: playlistCount)
            var playlistName: String = ""
            Database.shared().localPlaylistsDbQueue?.inDatabase { db in
                if let rs = db.executeQuery("SELECT playlist FROM localPlaylists WHERE md5 = ?", withArgumentsIn: [md5]) {
                    if rs.next() { playlistName = rs.string(forColumnIndex: 0) ?? "" }
                    rs.close()
                }
            }
            title = playlistName
        }

        tableView.backgroundColor = UIColor(named: "isubBackgroundColor")
        tableView.separatorStyle = .none
        tableView.rowHeight = Defines.rowHeight
        tableView.register(
            UniversalTableViewCell.self,
            forCellReuseIdentifier: UniversalTableViewCell.reuseId
        )
    }

    override func viewWillAppear(_ animated: Bool) {
        super.viewWillAppear(animated)

        // Now-playing button
        navigationItem.rightBarButtonItem = Music.shared().showPlayerIcon
            ? UIBarButtonItem(
                image: UIImage(systemName: Defines.musicNoteImageSystemName),
                style: .plain,
                target: self,
                action: #selector(nowPlayingAction(_:))
              )
            : nil

        switch source {
        case .local:
            reloadLocalSongs()

        case .server:
            if songs.isEmpty {
                fetchServerSongs()
            }
        }
    }

    override func viewWillDisappear(_ animated: Bool) {
        super.viewWillDisappear(animated)
        fetchTask?.cancel()
    }

    deinit {
        fetchTask?.cancel()
    }

    // MARK: - Data loading

    private func reloadLocalSongs() {
        guard case .local(let md5, _) = source else { return }

        // Re-read current count from DB so a push-back-then-return reflects edits
        var count: Int32 = 0
        Database.shared().localPlaylistsDbQueue?.inDatabase { db in
            if let rs = db.executeQuery("SELECT COUNT(*) FROM playlist\(md5)", withArgumentsIn: []) {
                if rs.next() { count = rs.int(forColumnIndex: 0) }
                rs.close()
            }
        }
        self.playlistCount = UInt(count)
        self.source = .local(md5: md5, count: UInt(count))
        let playlist = ISMSLocalPlaylist(name: title ?? "", md5: md5, count: UInt(count))
        songs = LocalPlaylistDAO().fetchSongs(in: playlist)

        updateHeader()
        tableView.reloadData()
    }

    private func fetchServerSongs() {
        guard case .server(let sp) = source else { return }

        ViewObjects.shared().showAlbumLoadingScreen(view, sender: self)
        tableView.isScrollEnabled = false

        fetchTask?.cancel()
        fetchTask = Task {
            do {
                let fetched = try await ServerPlaylistService().fetchSongs(in: sp)
                guard !Task.isCancelled else { return }
                await MainActor.run {
                    self.songs = fetched
                    self.playlistCount = UInt(fetched.count)
                    self.updateHeader()
                    self.tableView.reloadData()
                    self.tableView.isScrollEnabled = true
                    ViewObjects.shared().hideLoadingScreen()
                }
            } catch {
                guard !Task.isCancelled else { return }
                await MainActor.run {
                    self.tableView.isScrollEnabled = true
                    ViewObjects.shared().hideLoadingScreen()
                    if Settings.shared().isPopupsEnabled {
                        let msg = "There was an error loading the playlist.\n\n\(error.localizedDescription)"
                        let alert = UIAlertController(
                            title: "Error",
                            message: msg,
                            preferredStyle: .alert
                        )
                        alert.addAction(UIAlertAction(title: "OK", style: .cancel))
                        self.present(alert, animated: true)
                    }
                }
            }
        }
    }

    // MARK: - Header

    private func updateHeader() {
        let artIds: [String] = songs.prefix(4).compactMap { $0.coverArtId }

        let header = PlaylistHeaderView()
        header.playlistName = title
        header.songCount = songs.count
        header.coverArtIds = artIds
        header.onPlayAll = { [weak self] in self?.playAll() }
        header.onShuffle = { [weak self] in self?.shuffleAll() }

        // Wrap in a container so Auto Layout can size the tableHeaderView properly
        let container = UIView()
        container.translatesAutoresizingMaskIntoConstraints = false

        header.translatesAutoresizingMaskIntoConstraints = false
        container.addSubview(header)
        NSLayoutConstraint.activate([
            header.leadingAnchor.constraint(equalTo: container.leadingAnchor),
            header.trailingAnchor.constraint(equalTo: container.trailingAnchor),
            header.topAnchor.constraint(equalTo: container.topAnchor),
            header.heightAnchor.constraint(equalToConstant: PlaylistHeaderView.height),
            header.bottomAnchor.constraint(equalTo: container.bottomAnchor),
        ])

        tableView.tableHeaderView = container
        NSLayoutConstraint.activate([
            container.centerXAnchor.constraint(equalTo: tableView.centerXAnchor),
            container.widthAnchor.constraint(equalTo: tableView.widthAnchor),
            container.topAnchor.constraint(equalTo: tableView.topAnchor),
        ])
        container.layoutIfNeeded()
        // Force UITableView to adopt the header's intrinsic size
        tableView.tableHeaderView = tableView.tableHeaderView
    }

    // MARK: - Playback helpers

    /// Copies all songs from the in-memory `songs` array into the current playlist DB and plays from `position`.
    private func loadPlaylistAndPlay(at position: Int) {
        let settings = Settings.shared()
        let database = Database.shared()
        let jukebox = Jukebox.shared()
        let playQueue = PlayQueue.shared()

        if settings.isJukeboxEnabled {
            database.resetJukeboxPlaylist()
            jukebox.clearRemotePlaylist()
        } else {
            database.resetCurrentPlaylistDb()
        }

        playQueue.isShuffle = false

        let currTableName = settings.isJukeboxEnabled ? "jukeboxCurrentPlaylist" : "currentPlaylist"
        if let queue = database.currentPlaylistDbQueue {
            for song in songs {
                song.insertIntoTable(currTableName, inDatabaseQueue: queue)
            }
        }

        if settings.isJukeboxEnabled {
            jukebox.replacePlaylistWithLocal()
        }

        Music.shared().playSong(atPosition: position)
    }

    private func playAll() {
        loadPlaylistAndPlay(at: 0)
    }

    private func shuffleAll() {
        let settings = Settings.shared()
        let database = Database.shared()
        let jukebox = Jukebox.shared()

        if settings.isJukeboxEnabled {
            database.resetJukeboxPlaylist()
            jukebox.clearRemotePlaylist()
        } else {
            database.resetCurrentPlaylistDb()
        }

        PlayQueue.shared().isShuffle = true

        let currTableName = settings.isJukeboxEnabled ? "jukeboxCurrentPlaylist" : "currentPlaylist"
        if let queue = database.currentPlaylistDbQueue {
            for song in songs {
                song.insertIntoTable(currTableName, inDatabaseQueue: queue)
            }
        }

        if settings.isJukeboxEnabled {
            jukebox.replacePlaylistWithLocal()
        }

        Music.shared().playSong(atPosition: 0)
    }

    private func addAllToQueue() {
        for song in songs {
            song.addToCurrentPlaylistDbQueue()
        }
        SlidingNotification.showOnMainWindow(message: "All songs added to queue", duration: 1.5)
        HapticEngine.shared.success()
    }

    // MARK: - Actions

    @objc private func nowPlayingAction(_ sender: Any?) {
        let vc = PlayerViewController()
        vc.hidesBottomBarWhenPushed = true
        navigationController?.pushViewController(vc, animated: true)
    }

    // MARK: - Server playlist: Add Songs (placeholder, wired up in a future step)

    @objc private func addSongsAction(_ sender: Any?) {
        // TODO: Present song browser to add songs to this server playlist
    }
}

// MARK: - UITableViewDataSource

extension PlaylistSongsViewController {

    override func numberOfSections(in tableView: UITableView) -> Int { 1 }

    override func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        songs.count
    }

    override func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        let cell = tableView.dequeueReusableCell(
            withIdentifier: UniversalTableViewCell.reuseId,
            for: indexPath
        ) as! UniversalTableViewCell

        let song = songs[indexPath.row]
        cell.hideNumberLabel = false
        cell.hideCoverArt    = false
        cell.hideDurationLabel = false
        cell.hideSecondaryLabel = false
        cell.number = indexPath.row + 1
        cell.update(model: song)
        cell.backgroundColor = .clear

        if !song.isVideo {
            let isServer = { if case .server = self.source { return true }; return false }()
            if isServer {
                cell.contextMenuProvider = { [weak self] _ in
                    self?.serverSongContextMenu(song: song, index: indexPath.row)
                }
            } else {
                cell.configureSongContextMenu(song: song, presenter: self)
            }
        }

        return cell
    }
}

// MARK: - UITableViewDelegate

extension PlaylistSongsViewController {

    override func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
        tableView.deselectRow(at: indexPath, animated: true)
        ViewObjects.shared().showLoadingScreenOnMainWindow(withMessage: nil)
        // Defer slightly so the loading screen appears before we block the main thread
        DispatchQueue.main.asyncAfter(deadline: .now() + 0.05) { [weak self] in
            self?.loadPlaylistAndPlay(at: indexPath.row)
        }
    }

    override func tableView(
        _ tableView: UITableView,
        trailingSwipeActionsConfigurationForRowAt indexPath: IndexPath
    ) -> UISwipeActionsConfiguration? {
        let song = songs[indexPath.row]
        guard !song.isVideo else { return nil }
        return SwipeAction.downloadAndQueueConfig(model: song)
    }

    override func tableView(
        _ tableView: UITableView,
        contextMenuConfigurationForRowAt indexPath: IndexPath,
        point: CGPoint
    ) -> UIContextMenuConfiguration? {
        // Context menus are handled via UniversalTableViewCell's UIContextMenuInteraction;
        // returning nil here prevents double-triggering from the table view delegate.
        return nil
    }
}

// MARK: - Context menus

private extension PlaylistSongsViewController {

    /// Context menu for server-playlist rows (adds "Remove from Playlist" on top of the standard items).
    func serverSongContextMenu(song: Song, index: Int) -> UIMenu {
        let play = UIAction(
            title: "Play",
            image: UIImage(systemName: "play.fill")
        ) { [weak self] _ in
            guard let self else { return }
            ViewObjects.shared().showLoadingScreenOnMainWindow(withMessage: nil)
            DispatchQueue.main.asyncAfter(deadline: .now() + 0.05) {
                self.loadPlaylistAndPlay(at: index)
            }
        }

        let addToQueue = UIAction(
            title: "Add to Queue",
            image: UIImage(systemName: "text.badge.plus")
        ) { _ in
            song.addToCurrentPlaylistDbQueue()
            SlidingNotification.showOnMainWindow(message: "Added to queue", duration: 1.0)
            HapticEngine.shared.success()
        }

        let addToPlaylist = UIAction(
            title: "Add to Playlist…",
            image: UIImage(systemName: "music.note.list")
        ) { [weak self] _ in
            guard let self else { return }
            HapticEngine.shared.secondaryAction()
            let vc = AddToPlaylistViewController(song: song)
            let nav = UINavigationController(rootViewController: vc)
            if let sheet = nav.sheetPresentationController {
                sheet.detents = [.medium(), .large()]
                sheet.prefersGrabberVisible = true
            }
            self.present(nav, animated: true)
        }

        let remove = UIAction(
            title: "Remove from Playlist",
            image: UIImage(systemName: "minus.circle"),
            attributes: .destructive
        ) { [weak self] _ in
            guard let self, case .server(let sp) = self.source else { return }
            Task { [weak self] in
                guard let self else { return }
                do {
                    try await ServerPlaylistService().removeSongs(at: [index], from: sp)
                    await MainActor.run {
                        self.songs.remove(at: index)
                        self.tableView.deleteRows(at: [IndexPath(row: index, section: 0)], with: .automatic)
                        self.updateHeader()
                    }
                } catch {
                    await MainActor.run {
                        SlidingNotification.showOnMainWindow(message: "Failed to remove song")
                    }
                }
            }
        }

        return UIMenu(title: song.title ?? "", children: [play, addToQueue, addToPlaylist, remove])
    }
}
